libdnf/comps/environment/query.cpp

```diff
@@ -63,25 +65,29 @@ EnvironmentQuery::EnvironmentQuery(const BaseWeakPtr & base) : base(base) {
             continue;
         }
 
+        repoid = solvable->repo->name;
+
         // Add installed environments directly, because there is only one solvable for each
-        if (strcmp(solvable->repo->name, "@System") == 0) {
+        if (strcmp(repoid, "@System") == 0) {
             Environment environment(base);
             environment.environment_ids.push_back(EnvironmentId(solvable_id));
             add(environment);
         } else {
             // Create map of available environments:
             // for each environmentid (SOLVABLE_NAME), list all corresponding solvable_ids
-            // TODO(pkratoch): Sort solvable_ids for each environmentid according to repo id
             available_map[solvable_name_pair.second].insert(
-                available_map[solvable_name_pair.second].begin(), solvable_id);
+                available_map[solvable_name_pair.second].end(), std::pair<Id, std::string_view>(solvable_id, repoid));
         }
     }
 
     // Create environments based on the available_map
-    for (const auto & item : available_map) {
+    for (auto item : available_map) {
         Environment environment(base);
-        for (Id solvable_id : item.second) {
-            environment.environment_ids.push_back(EnvironmentId(solvable_id));
+        // Sort the vector of (solvable_id, repoid) pairs by repoid
+        std::sort(item.second.begin(), item.second.end(), [](auto a, auto b) { return a.second > b.second; });
```

Comment:
Since you're storing the repoid in the vector now (which makes sense, I originally thought you'd pull the repoid from the solvables during the comparison but this is much more effective), if you put the repoid as the first item in the vector, you wouldn't need the lambda at all, it would work automatically via operators on the pair and string_view.

With the lambda, you should again use `auto &` (or `const auto &`), otherwise you're making copies just for the comparison.

Reply:
It still needs to be reversed, so that the behavior is the same as in dnf 4, so I used `std::greater<>()`. Maybe I could instead iterate it in reverse afterwords, but I don't know which method is more efficient.

Reply:
Isn't it less intuitive that way? Do we need to preserve compatibility with this? I'd just do the "natural" ordering unless I'm missing something...

Reply:
When getting for example group name, the `group_ids` are searched in order and first that is found is used. So, if ordered naturally, it would take name from repo `fedora` instead of `updates`. I don't know how big problem it would cause if we changed the order. Probably the comps don't change that often, but I'd rather keep it as it is.

Reply:
Ok, I see. Do we have this documented somewhere in the comments? (sorry I don't remember and don't want to go through everything again)

I think we should have this documented in some appropriate place in the code in detail, including the "coincidental" interaction with the `fedora` and `updates` repos, this is an important information that's not obvious.

Reply:
Good idea, I added it to the declaration of group_ids and environment_ids.

Reply:
Perfect, thank you :slightly_smiling_face: